v2/api_error.go

```diff
@@ -0,0 +1,213 @@
+			strings.Join(f, " "), strings.Join(desc, " ")))
+	}
+
+	if a.details.PreconditionFailure != nil {
+		v := a.details.PreconditionFailure.GetViolations()
+		var t []string
+		var s []string
+		var desc []string
+		for _, x := range v {
+			t = append(t, x.GetType())
+			s = append(s, x.GetSubject())
+			desc = append(desc, x.GetDescription())
+		}
+		d.WriteString(fmt.Sprintf("error details: name = PreconditionFailure type = %s subj = %s desc = %s", strings.Join(t, " "),
+			strings.Join(s, " "), strings.Join(desc, " ")))
+	}
+
+	if a.details.QuotaFailure != nil {
+		v := a.details.QuotaFailure.GetViolations()
+		var s []string
+		var desc []string
+		for _, x := range v {
+			s = append(s, x.GetSubject())
+			desc = append(desc, x.GetDescription())
+		}
+		d.WriteString(fmt.Sprintf("error details: name = QuotaFailure subj = %s desc = %s",
+			strings.Join(s, " "), strings.Join(desc, " ")))
+	}
+
+	if a.details.RequestInfo != nil {
+		d.WriteString(fmt.Sprintf("error details: name = RequestInfo id = %s data = %s",
+			a.details.RequestInfo.GetRequestId(), a.details.RequestInfo.GetServingData()))
+	}
+
+	if a.details.ResourceInfo != nil {
+		d.WriteString(fmt.Sprintf("error details: name = ResourceInfo type = %s resourcename = %s owner = %s desc = %s",
+			a.details.ResourceInfo.GetResourceType(), a.details.ResourceInfo.GetResourceName(),
+			a.details.ResourceInfo.GetOwner(), a.details.ResourceInfo.GetDescription()))
+
+	}
+	if a.details.RetryInfo != nil {
+		d.WriteString(fmt.Sprintf("error details: retry in %s", a.details.RetryInfo.GetRetryDelay().AsDuration()))
+
+	}
+	if a.details.Unknown != nil {
+		var s []string
+		for _, x := range a.details.Unknown {
+			s = append(s, fmt.Sprintf("%v", x))
+		}
+		d.WriteString(fmt.Sprintf("error details: name = Unknown  desc = %s", strings.Join(s, " ")))
+
+	}
+	if a.details.DebugInfo != nil {
+		d.WriteString(fmt.Sprintf("error details: name = DebugInfo detail = %s stack = %s", a.details.DebugInfo.GetDetail(),
+			strings.Join(a.details.DebugInfo.GetStackEntries(), " ")))
+	}
+	if a.details.Help != nil {
+		var desc []string
+		var url []string
+		for _, x := range a.details.Help.Links {
+			desc = append(desc, x.GetDescription())
+			url = append(url, x.GetUrl())
+		}
+		d.WriteString(fmt.Sprintf("error details: name = Help desc = %s url = %s",
+			strings.Join(desc, " "), strings.Join(url, " ")))
+	}
+	if a.details.LocalizedMessage != nil {
+		d.WriteString(fmt.Sprintf("error details: name = LocalizedMessge locale = %s msg = %s",
+			a.details.LocalizedMessage.GetLocale(), a.details.LocalizedMessage.GetMessage()))
+	}
+	return strings.TrimSpace(d.String())
+}
+
+// GRPCStatus extracts underlying gRPC status in a APIError
+func (a *APIError) GRPCStatus() *status.Status {
+	return a.status
+}
+
+// FromError builds an APIError from an error
+func FromError(err error) (*APIError, bool) {
+	if err == nil {
+		return nil, false
+	}
+	st, ok := status.FromError(err)
+	if !ok {
```

Comment:
I wonder if this method should not fail, but if it could not parse a Status just return the raw error passed in? Also maybe document that this method is to parse status error.

Reply:
When we try to wrap the error in invoke.go, if FromError returns false, the original error is returned but would it be better to still return the original error here? @codyoss

Reply:
We need to return an `*APIError` not an `error` so we cannot return the original error. Just like `status.FromError`, it will return `nil, false` if it is unable to parse out a `status.Status`.

If we had a `New(err error) error` of some sort, and we couldn't parse the `err` into an `APIError`, I'd say it should return the original `error`.